irohad/ordering/impl/on_demand_ordering_service_impl.cpp

```diff
@@ -4,25 +4,147 @@
+    return clone(*proposal->second);
+  } else {
+    return boost::none;
+  }
+}
+
+// ---------------------------------| Private |---------------------------------
+
+void OnDemandOrderingServiceImpl::packNextProposal(
+    RoundOutput outcome, const transport::RoundType &last_round) {
+  log_->info("pack next proposal...size of queue = {}",
+             current_proposal_.second.unsafe_size());
+  if (not current_proposal_.second.empty()) {
+    proposal_map_.insert(
+        std::make_pair(current_proposal_.first, emitProposal()));
+    log_->info("packNextProposal: data has been fetched");
+  }
+
+  round_queue_.push(current_proposal_.first);
+  log_->info("packNextProposal: pushed in queue");
+
+  auto current_round = current_proposal_.first;
+  decltype(current_round) next_round;
+  switch (outcome) {
+    case RoundOutput::SUCCESSFUL:
+      next_round = std::make_pair(current_round.first + 1, kFirstRound);
+      break;
+    case RoundOutput::REJECT:
+      next_round =
+          std::make_pair(current_round.first, current_round.second + 1);
+      break;
+  }
+
+  log_->info("nextRound is: [{}, {}]", next_round.first, next_round.second);
+
+  current_proposal_.first = next_round;
+  current_proposal_.second.clear();  // operator = of tbb::queue is closed
 }
 
-boost::optional<std::shared_ptr<shared_model::interface::Proposal>>
-OnDemandOrderingServiceImpl::onRequestProposal(uint64_t round) {
-  throw std::logic_error("Function not yet implemented");
+OnDemandOrderingServiceImpl::ProposalType
+OnDemandOrderingServiceImpl::emitProposal() {
+  iroha::protocol::Proposal proto_proposal;
+  proto_proposal.set_height(current_proposal_.first.first);
+  proto_proposal.set_created_time(iroha::time::now());
+  log_->info("Mutable proposal generation");
+
+  TransactionType current_tx;
+  using ProtoTxType = shared_model::proto::Transaction;
+  std::vector<TransactionType> collection;
+
+  // outer method should guarantee availability of at least one transaction in
+  // queue, also, code shouldn't fetch all transactions from queue. The rest
+  // will be lost.
+  while (current_proposal_.second.try_pop(current_tx)
+         and collection.size() < transaction_limit_) {
+    collection.emplace_back(std::move(current_tx));
+  }
+  log_->info("Number of transaction in proposal  = {}", collection.size());
+  auto proto_txes = collection | boost::adaptors::transformed([](auto &tx) {
+                      return static_cast<const ProtoTxType &>(*tx);
+                    });
+  boost::for_each(proto_txes, [&proto_proposal](auto &&proto_tx) {
+    *(proto_proposal.add_transactions()) = std::move(proto_tx.getTransport());
+  });
+
+  return std::make_unique<shared_model::proto::Proposal>(
+      std::move(proto_proposal));
+}
+
+void OnDemandOrderingServiceImpl::tryErase() {
+  if (round_queue_.size() < number_of_proposals_) {
+    return;
+  }
+
+  if (round_queue_.empty() and round_queue_.front().second == kFirstRound) {
```

Comment:
This will lead to undefined behavior if `round_queue_` is empty, because it will start evaluating the second expression which accesses `front`. Please fix, add comments for conditionals and simplify if possible.